tools/calyx-pass/src/pass_explorer.rs

```diff
@@ -0,0 +1,302 @@
+            line.trim_start_matches(|c: char| !c.is_alphanumeric())
+                .to_string()
+        })
+        .collect();
+
+        assert!(!passes.is_empty(), "Unknown pass alias '{}'", pass_alias);
+
+        let mut dest_path = PathBuf::from(work_dir.path());
+        dest_path.push(SOURCE_FILE_NAME);
+        fs::copy(input_file, dest_path.clone())?;
+
+        let mut new_self = Self {
+            work_dir,
+            calyx_exec,
+            breakpoint,
+            passes,
+            passes_applied: vec![],
+            index: 0,
+            file_exists: HashSet::new(),
+        };
+
+        if let Some(breakpoint) = new_self.breakpoint.clone() {
+            assert!(new_self.passes.contains(&breakpoint.pass));
+            while new_self.incoming_pass().expect("There is at least one pass by our prior assertion and we also must encounter the breakpoint").ne(&breakpoint.pass) {
+                if !breakpoint.skip.contains(&new_self.incoming_pass().unwrap()) {
+                new_self.ensure_inc_file_exists()?;
+                new_self.accept()?;
+                } else {
+                    new_self.skip()?;
+                }
+            }
+        }
+
+        Ok(new_self)
+    }
+
+    /// The pass most recently applied, if one exists.
+    pub fn last_pass(&self) -> Option<String> {
+        self.passes_applied
+            .last()
+            .map(|index| self.passes[*index].clone())
+    }
+
+    /// The pass staged to be applied, if one exists.
+    pub fn incoming_pass(&self) -> Option<String> {
+        self.passes.get(self.index as usize).cloned()
+    }
+
+    /// An association of each pass with its current exploration status.
+    pub fn current_pass_application(
+        &self,
+    ) -> Vec<(String, PassApplicationStatus)> {
+        let mut result = vec![];
+        let mut walk_applied = 0;
+        for i in 0..self.passes.len() {
+            let pass = self.passes[i].clone();
+            if walk_applied < self.passes_applied.len()
+                && i == self.passes_applied[walk_applied]
+            {
+                result.push((pass, PassApplicationStatus::Applied));
+                walk_applied += 1;
+            } else if i < (self.index as usize) {
+                result.push((pass, PassApplicationStatus::Skipped));
+            } else if i == (self.index as usize) {
+                result.push((pass, PassApplicationStatus::Incoming));
+            } else {
+                result.push((pass, PassApplicationStatus::Future));
+            }
+        }
+        result
+    }
+
+    /// Produces a printable diff showing how the
+    /// [`PassExplorer::incoming_pass`] will transform the current file state.
+    pub fn review(
+        &mut self,
+        component: Option<String>,
+    ) -> std::io::Result<Option<String>> {
+        self.ensure_inc_file_exists()?;
+        let mut last_file_content = fs::read_to_string(self.last_file())
+            .expect("Could not read the last file");
```

Comment:
`expect` will crash the tool instead of gracefully showing an error. Why not bubble up the error message?